src/EFCore.PG/Storage/Internal/Mapping/NpgsqlImmutableHstoreTypeMapping.cs

```diff
@@ -0,0 +1,44 @@
+using System.Collections.Immutable;
+using Microsoft.EntityFrameworkCore.ChangeTracking;
+using Microsoft.EntityFrameworkCore.Storage;
+
+namespace Npgsql.EntityFrameworkCore.PostgreSQL.Storage.Internal.Mapping
+{
+    /// <summary>
+    /// The type mapping for the PostgreSQL hstore type to <see cref="ImmutableDictionary"/>.
+    /// </summary>
+    /// <remarks>
+    /// See: https://www.postgresql.org/docs/current/static/hstore.html
+    /// </remarks>
+    public class NpgsqlImmutableHstoreTypeMapping : NpgsqlHstoreTypeMapping
+    {
+        static readonly HstoreComparer ComparerInstance = new HstoreComparer();
+
+        /// <summary>
+        /// Constructs an instance of the <see cref="NpgsqlImmutableHstoreTypeMapping"/> class.
+        /// </summary>
+        public NpgsqlImmutableHstoreTypeMapping()
+            : base(
+                new RelationalTypeMappingParameters(
+                    new CoreTypeMappingParameters(typeof(ImmutableDictionary<string, string>), null, ComparerInstance),
+                    "hstore"
+                )) {}
+
+        protected NpgsqlImmutableHstoreTypeMapping(RelationalTypeMappingParameters parameters)
+            : base(parameters) {}
+
+        protected override RelationalTypeMapping Clone(RelationalTypeMappingParameters parameters)
+            => new NpgsqlImmutableHstoreTypeMapping(parameters);
+
+        sealed class HstoreComparer : ValueComparer<ImmutableDictionary<string, string>>
```

Comment:
/cc @ajcvickers for a really nice use of a .NET immutable type for version tracking purposes. tl;dr PG has a type called "hstore", which corresponds to .NET Dictionary<string, string?>. This PR also allows you to map hstore to .NET ImmutableDictionary<string, string?>, which (hopefully) efficiently creates copies on updates. This means we can do simple reference comparisons, and skip actually snapshotting altogether.

Could be interesting to think about for other types (e.g. ImmutableArray for mapping blobs/varbinary??).